[DOCS] Clarify testing section (#7632)
ethereum/solidity: docs/contributing.rst

Running the compiler tests
==========================

Prerequisites
-------------

Some tests require the `evmone <https://github.com/ethereum/evmone/releases>`_
library, others require `libz3 <https://github.com/Z3Prover/z3>`_. The test script
tries to discover the location of the ``evmone`` library, which can be located in

> the current directory, or the ``deps`` folder in the project top level. The required

Reviewer: ```suggestion
the current directory, installed on the system level, or the ``deps`` folder in the project top level. The required
```

including those bundled into the `Boost C++ Test Framework <https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/index.html>`_ application ``soltest`` (or its wrapper ``scripts/soltest.sh``),
as well as command line tests and compilation tests.

The test system automatically tries try to discover the location of the ``evmone`` library
starting from the current directory. The required file is called ``libevmone.so`` on Linux systems,
``evmone.dll`` on Windows systems and ``libevmone.dylib`` on macOS. If it is not found, tests that
use it are skipped. These tests are ``libsolididty/semanticTests``, ``libsolidity/GasCosts``,
``libsolidity/SolidityEndToEndTest``, part of the soltest suite. To run all tests, download the library from
`GitHub <https://github.com/ethereum/evmone/releases/tag/v0.3.0>`_
and place it in the project root path or inside the ``deps`` folder.

If the ``libz3`` library is not installed on your system, you should disable the
SMT tests using the ``--no-smt`` option when running ``./scripts/tests.sh`` or
``./scripts/soltest.sh``. These tests are ``libsolidity/smtCheckerTests`` and ``libsolidity/smtCheckerTestsJSON``.

.. csv-table:: The full test suite

>    :file: ./contributing/tests-list.csv

Reviewer 2: I don't think this is very helpful. What is the information we convey here? This list will be outdated very soon and I don't think we want to have the overhead of updating it.

Author: I have no knowledge of what tests are changing and when, and this PR has sat here for a long time. I'm reacting to the Zeppelin audit and comments from contributors that they don't understand what tests we have and what they do, hence a reference might be useful to them. Otherwise there are a lot of tests that are supposed to pass before someone can make a contribution, but most people (including some core team hires) take time to understand what they are. So a reference seems helpful to all of these people. If things change, that's fine, but a reference is still useful.

If you render the RsT, or look at the csv it lists what tests run as part of which test suite and what they are testing.

I am open to other suggestions (like maybe commenting the test files instead), but as this has come up a bit, I feel it's valuable to contributors.

Reviewer: @chriseth opinion after @ChrisChinchilla 's reply?

Reviewer: @bshastry The error you pointed out was actually a bug in the SMTChecker docs, should be fixed by https://github.com/ethereum/solidity/pull/8470. The `--no-smt` flag works for `soltest` and `isoltest`, but doesn't relay anything to the command line tests.

Reviewer 3: @leonardoalt Are you referring to the comment about SMT checker not being disabled by the bash tests script (https://github.com/ethereum/solidity/pull/7632#discussion_r358373498) or the other one about string comparison in bash (https://github.com/ethereum/solidity/pull/7632#discussion_r358379555)?

Reviewer: First one

Reviewer 3: > The --no-smt flag works for soltest and isoltest, but doesn't relay anything to the command line tests.

I'm a bit confused. Do you mean `$ ./scripts/tests.sh --no-smt` will work as intended post #8470 ?

Reviewer: `--no-smt` has always worked when relayed to `soltest` and `isoltest`. What didn't work was the command line tests when no solver is available. That is fixed by #8470

Reviewer 3: Okay, thanks @leonardoalt 

In that case, please ignore my earlier comment

Reviewer: @chriseth The `csv` file was removed

tries to discover the location of the ``evmone`` library, which can be located in
the current directory, or the ``deps`` folder in the project top level. The required
file is called ``libevmone.so`` on Linux systems, ``evmone.dll`` on Windows systems
and ``libevmone.dylib`` on macOS.

Running the tests
-----------------

Solidity includes different types of tests, most of them bundled into the
`Boost C++ Test Framework <https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/index.html>`_ application ``soltest``.
Running ``build/test/soltest`` or its wrapper ``scripts/soltest.sh`` is sufficient for most changes.

The ``./scripts/tests.sh`` script executes most Solidity tests automatically,
including those bundled into the `Boost C++ Test Framework <https://www.boost.org/doc/libs/1_69_0/libs/test/doc/html/index.html>`_ application ``soltest`` (or its wrapper ``scripts/soltest.sh``),
as well as command line tests and compilation tests.

The test system automatically tries try to discover the location of the ``evmone`` library
starting from the current directory. The required file is called ``libevmone.so`` on Linux systems,

> ``evmone.dll`` on Windows systems and ``libevmone.dylib`` on MacOS. If it is not found, the relevant tests

Reviewer 3: ```suggestion
 ``evmone.dll`` on Windows systems and ``libevmone.dylib`` on MacOS. If it is not found, the following tests (part of the soltest suite) are skipped:
```

Author: Ahh, I see what you mean. OK, I won't accept this suggestion, but move things around a bit to reflect some of what you suggest.

Reviewer 3: Sure, it was just a suggestion :-)

> are skipped. To run all tests, download the library from

Reviewer 3: ```suggestion
``libsolididty/semanticTests``, ``libsolidity/GasCosts``, and ``libsolidity/SolidityEndToEndTest``. To run all tests, download the library from
```

`Github <https://github.com/ethereum/evmone/releases/tag/v0.3.0>`_
and either place it in the project root path or inside the ``deps`` folder.

> If the ``evmone`` library is not found, tests that use it are skipped. These tests

Reviewer 3: This sentence is a duplicate of `If it is not found, the relevant tests ￼are skipped.` a few lines above.

Author: I had a merge conflict before last push, so maybe I missed something in the tidy up

Author: Actually, this isn't duplication, the first paragraph says how to get evmone, the other says what tests are skipped if it can't be found. To make it clearer, I merged the paragraphs

> If the ``evmone`` library is not found, tests that use it are skipped. These tests

Reviewer 3: ```suggestion
```

> are ``libsolididty/semanticTests``, ``libsolidity/GasCosts``, ``libsolidity/SolidityEndToEndTest``

Reviewer 3: ```suggestion
```

> part of the soltest suite.

Reviewer 3: ```suggestion
```

If the ``libz3`` library is not installed on your system, you should disable the

> SMT tests using the ``--no-smt`` option when running ``./scripts/tests.sh`` or

Reviewer 3: I tested this but failed to disable SMT tests. I did

```
$ ./scripts/tests.sh --no-smt
...
test_0571770b984517916a9a795a2b5f8a37ce1c20f5eee047a30f3e5315718c32d4_s
ecurity_considerations_rst.sol
Unexpected compilation result:       
Expected failure: 0 - Expected warning / error output: 0
Was failure: 0

Warning: BMC analysis was not possible since no integrated SMT solver (
Z3 or CVC4) was found.
While calling:
...
```

Are you sure this flag works as intended?

Author: Adding the possibility is in this branch, did you try when checked out / built from this branch?

Reviewer 3: Yes, I remember checking out the branch before I ran the command in my previous comment in this thread.

Author: Hmm, odd, it works for me. You generally have to rebuild everything to get tests correct, but I don't think they are affected by this… I can check on Linux I guess, but this just copied syntax I saw elsewhere in our bash scripts

Reviewer 3: Okay, I'll check again (my system is Linux anyway).

``./scripts/soltest.sh``. These tests are ``libsolidity/smtCheckerTests`` and ``libsolidity/smtCheckerTestsJSON``.

.. csv-table:: The full test suite
   :file: ./contributing/tests-list.csv
   :widths: 40, 60
   :header-rows: 1

.. note ::

    To get a list of all unit tests run by Soltest, run ``./build/test/soltest --list_content=HRF``.

For quicker results you can run a subset of, or specific tests.

To run a subset of tests, you can use filters:
``./scripts/soltest.sh -t TestSuite/TestName``,
where ``TestName`` can be a wildcard ``*``.

> Or, for example, to run all the tests for the disambiguator:

Reviewer 3: ```suggestion
Or, for example, to run all the tests for the yul disambiguator:
```
